proof/refine/ARM/Tcb_R.thy

```diff
@@ -3547,14 +3549,66 @@ lemma decode_set_space_corres:
    apply fastforce
   apply (frule list_all2_lengthD)
   apply (clarsimp split: list.split)
-  done *)
+  done
+
+lemma decodeSetSpace_corres:
+ "\<lbrakk>cap_relation cap cap';
+   list_all2 (\<lambda>(c, sl) (c', sl'). cap_relation c c' \<and> sl' = cte_map sl) extras extras';
+   is_thread_cap cap\<rbrakk> \<Longrightarrow>
+  corres (ser \<oplus> tcbinv_relation)
+         (invs and valid_cap cap and (\<lambda>s. \<forall>x \<in> set extras. cte_at (snd x) s))
+         (invs' and valid_cap' cap' and (\<lambda>s. \<forall>x \<in> set extras'. cte_at' (snd x) s))
+      (decode_set_space args cap slot extras)
+      (decodeSetSpace args cap' (cte_map slot) extras')"
+  apply (simp add: decode_set_space_def decodeSetSpace_def check_handler_ep_def unlessE_whenE)
+  apply (cases "\<not> (2 \<le> length args \<and> 3 \<le> length extras')")
+   apply (clarsimp dest!: list_all2_lengthD split: list.split)
+   apply fastforce
+  apply (clarsimp simp: val_le_length_Cons list_all2_Cons2
+              split del: if_split)
+  apply (rule corres_guard_imp)
+    apply (rule corres_splitEE[OF _ decodeCVSpace_corres])
+         apply (rename_tac abs_space conc_space)
+         apply clarsimp
+         apply (intro conjI impI; simp add: cap_rel_valid_fh)
+         apply (prop_tac "newroot_rel (tc_new_croot abs_space) (tcCapsCRoot conc_space)")
+          apply (case_tac abs_space; clarsimp)
+         apply (prop_tac "newroot_rel (tc_new_vroot abs_space) (tcCapsVRoot conc_space)")
+          apply (case_tac abs_space; clarsimp)
+         apply (rule corres_returnOkTT)
+         apply (clarsimp simp: returnOk_def newroot_rel_def is_cap_simps list_all2_conv_all_nth)
+        apply blast
+       apply fastforce+
+     apply wp+
+   apply fastforce+
+  done
+
+lemma decodeCVSpace_wf[wp]:
+  "\<lbrace>invs' and tcb_at' t and ex_nonz_cap_to' t and cte_at' slot
+      and (\<lambda>s. \<forall>x \<in> set extras. s \<turnstile>' fst x \<and> cte_at' (snd x) s \<and> t \<noteq> snd x \<and> t + 16 \<noteq> snd x)\<rbrace>
+     decodeCVSpace args (ThreadCap t) slot extras
+   \<lbrace>tcb_inv_wf'\<rbrace>,-"
+  apply (simp       add: decodeCVSpace_def Let_def split_def
```

Comment:
nitpick: check indentation in this lemma. Seems a little odd, but maybe it's just a style that you prefer.